apps/react-playground/src/app/Widgets/MessagingPrototype/services/messaging-app.service.tsx

```diff
@@ -0,0 +1,163 @@
+import Deso from 'deso-protocol';
+import { DerivedPrivateUserInfo } from 'deso-protocol-types';
+import {
+  DecryptedResponse,
+  USER_TO_SEND_MESSAGE_TO_1,
+} from '../consts/constants';
+import {
+  authorizeDerivedKey,
+  generateDefaultKey,
+  getEncryptedMessages,
+  login,
+  requestDerivedKey,
+} from './messaging.service';
+import {
+  getDerivedKeyResponse,
+  setDefaultKey,
+  setDerivedKeyResponse,
+} from './store';
+import { delay } from './utils';
+export const getConversationsMapAndPublicKeyToUsernameMapping = async (
+  deso: Deso,
+  derivedResponse: Partial<DerivedPrivateUserInfo>
+): Promise<{
+  conversationMap: DecryptedResponse;
+  publicKeyToUsername: { [key: string]: string };
+}> => {
+  if (!derivedResponse) {
+    alert('no derived response found');
+    return { conversationMap: {}, publicKeyToUsername: {} };
+  }
+  const messages = await getEncryptedMessages(deso);
+  const decryptedMessages = await deso.utils.decryptMessagesV3(
+    messages,
+    derivedResponse.messagingPrivateKey as string
+  );
+  const conversationMap: { [key: string]: any[] } = {};
+  const publicKeyToUsername: { [key: string]: string } = {};
+  const userKey = deso.identity.getUserKey();
+  Object.keys(decryptedMessages)?.forEach(async (key: string) => {
+    decryptedMessages[key].forEach((message) => {
+      const otherUsersKey =
+        userKey === message.RecipientPublicKeyBase58Check
+          ? message.SenderPublicKeyBase58Check
+          : message.RecipientPublicKeyBase58Check;
+      if (!conversationMap[otherUsersKey]) {
+        conversationMap[otherUsersKey] = [];
+      }
+      publicKeyToUsername[key] =
+        messages.PublicKeyToProfileEntry[key].Username ?? null;
```

Comment:
```suggestion
        messages.PublicKeyToProfileEntry[key]?.Username ?? null;
```
Can you do that? I think we'll have some error for anonymous public keys.